src/main/java/uk/gov/ch/oauth/Oauth2.java

```diff
@@ -111,11 +111,13 @@ public boolean oauth2VerifyNonce(final String nonce) {
      *
      * @return The Nonce String from within the session or null if not found.
      */
-    private String getSessionNonce() {
+    private String getSessionNonce(Session chSession) {
         String oauth2Nonce = null;
         try {
-            final Map<String, Object> data = sessionService.getSessionDataFromContext();
+            final Map<String, Object> data = chSession.getData();
             oauth2Nonce = (String) data.getOrDefault(SessionKeys.NONCE.getKey(), null);
+            data.put(SessionKeys.NONCE.getKey(), "");
+            LOGGER.debug("Verify nonce against expected nonce from session");
```

Comment:
For clarity, it might be better to have a `destroyNonce()` method that you call here. That gives a better separation of concerns.